### PR TITLE
clean up after running repair tests

### PR DESCRIPTION
**Pull request by cstyan**
I noticed that some files are changed and others are created when running tsdb tests
Signed-off-by: Callum Styan <callumstyan@gmail.com>

### CODE REVIEW

**Comment by krasi-georgiev** [comment]
Yeah I noticed that too, but I think the problem is with the test itself. It should revert the state after completing. Otherwise it doesn't allow running it twice and confuses git as well

I also  prefer to use  `.git/info/exclude` for such edge use-cases and other user specific workflows.

**Comment by cstyan** [comment]
@krasi-georgiev updated so that we're cleaning up after the repair tests, reverted the gitignore changes. It's a bit ugly atm but I need to sleep. I'll clean it up it in the morning.

**Comment by krasi-georgiev** [comment]
thanks , ping me when ready and I will review again.

**Comment by cstyan** [comment]
@krasi-georgiev ping :)

**Comment by krasi-georgiev** [comment]
I just had a look at the test I did for another PR and here is a lot easyer way to achieve the same:

```
// Create and work on the db copy so we can run the test many times.
tmpdir := "testdata/repair_index_chunks/snapshot"
defer os.RemoveAll(tmpdir)
testutil.Ok(t, db.Snapshot(tmpdir, false))
dbCopy, err := Open(tmpdir, nil, nil, DefaultOptions)
testutil.Ok(t, err)
```

**Comment by cstyan** [comment]
@krasi-georgiev yeah I tried that initially, we need tombstones in order to create a snapshot. I don't we want to deal with generating tombstones for this test.

**Comment by krasi-georgiev** [comment]
even if you need to put an empty tombstones file it would be a much cleaner solution without requiring so many lines of additional code.

**Comment by cstyan** [comment]
if all we needed was an empty tombstones file I would have just done that :) let me have another look

**Comment by cstyan** [comment]
@krasi-georgiev we can't use snapshot here anyways, it's the call to `db, err := Open("testdata/repair_index_version/", nil, nil, DefaultOptions)` that actually modifies the existing files.

**Comment by krasi-georgiev** [comment]
I made it work using:
```
bl := Block{
   meta: BlockMeta{
	ULID: ulid.ULID{},
   },
   dir: "testdata/repair_index_version/01BZJ9WJQPWHGNC2W4J9TA62KC",
}
copy(bl.meta.ULID[:], bl.dir)

tmpdir := "testdata/repair_index_chunks/snapshot"
testutil.Ok(t, bl.Snapshot(tmpdir))
```

I think we would need to do this quite often form now on so  why don't we make some more generic func which takes filepaths as a param and recursively makes hard links using `os.Link` and put it in:
`fileutil/fileutil.go`

Something like:
```
func CopyFiles( files ...string, destDir string){
  //  iterate and os.Link all files
}
```
than we can have: `CopyFile(ReadDir(dirpath),destDir)`

if you don't have time for the extra effort let me know and I can take over.

**Comment by cstyan** [comment]
Thanks @krasi-georgiev, I got it working via the block snapshotting. 

Unfortunately I don't understand this generic copy files helper, how do hardlinks help us here?

**Comment by krasi-georgiev** [comment]
look at the snapshot implementation it uses hard links instead of copying files.

**Comment by krasi-georgiev** [comment]
hardlinks is essentially like copying , but without even opening the files.

**Comment by cstyan** [comment]
Right, but it's linked to the same underlying data on disk. So modifying the file at the hardlink name modifies the file at the original name, because they're essentially the same file.

Looking at the snapshot code, I don't understand how using the hardlinks gets us a copy of the DB files that we can modify without making modifications to the originals.

**Comment by krasi-georgiev** [comment]
`CloneFiles` is another name that comes to mind.

**Comment by krasi-georgiev** [comment]
ooh yeah I missunderstood the hardlinks a bit. 
Lets use the original idea with  `os.Open/io.Copy`, but just make the func something like: `CopyFile(ReadDir(dirpath),destDir)` as I have another test that would use it right away.

File: repair_test.go [hunk 1]
```diff
@@ -1,16 +1,43 @@
...
 	"testing"
 
 	"github.com/prometheus/tsdb/chunks"
+	"github.com/prometheus/tsdb/fileutil"
 
 	"github.com/prometheus/tsdb/index"
 	"github.com/prometheus/tsdb/labels"
 )
 
+func copyFiles(tmpIndex, tmpMeta *os.File, dir string) error {
```
Comment (krasi-georgiev) [inline repair_test.go]: don't need this anymore.

**Comment by cstyan** [comment]
`fileutil.ReadDir` just returns the names as strings, we'd need the path to the files. Also we need to handle the case where some contents of `srcDir` are directories even if we don't also want to recursively copy from sub-directories.

If `func CopyFiles( files ...string, destDir string)` works better for your use case I can make that change, but I don't think passing `fileutil.ReadDir(srcDir)` is useful unless srcdir happens to also be `os.Getwd()`.

**Comment by krasi-georgiev** [comment]
```
// Reads files in the src directory and all sub directories recursively. 
func ReadDirs(dir) map[path]name, error{
  recursively walk all folders including the root dir and build the  `map[path]name` using the `func ReadDir(...)`
}

CopyFiles(ReadDirs(srcDir),dstDir)
```

something like this should do the trick. 

Than I will try if it will work for my usecase as well.

**Comment by cstyan** [comment]
@krasi-georgiev imo it's easier if `ReadDirs` returns `[]string`, but how do you want it to handle empty directories? For example for this repair index test, we need the `chunks` dir to exist in the copy of the tsdb, but we just create an empty db in the real dir before copying it. I can just add another mkdir in the copy, or make `ReadDirs` and `CopyFiles` handle empty dirs as well.

**Comment by krasi-georgiev** [comment]
I would say lets create dirs regardless of whether they will include any files. 
This solves our use case so it is a good start. We can always update later if needed.

**Comment by krasi-georgiev** [comment]
for `[]string` vs `map[path]name` just do whichever would require less lines of code.

File: fileutil/fileutil.go [hunk 2]
```diff
@@ -5,11 +5,101 @@
...
+	}
+
+	if copied, err := io.Copy(temp, f); err != nil || copied != stat.Size() {
+		return err
+	}
+	if err = os.Rename(temp.Name(), dest); err != nil {
+		return err
+	}
+	return nil
+}
```
Comment (krasi-georgiev) [inline fileutil/fileutil.go]: why not something more simple like:

```
 data, err := ioutil.ReadFile(src)
 if err != nil {
     return err
  }

 err = ioutil.WriteFile(dst, data, 0644)
 if err != nil {
      return err
 }
```

File: fileutil/fileutil.go [hunk 3]
```diff
@@ -5,11 +5,102 @@
...
+	// path/copy/to/file, this function should return /copy
+	dir, _ := filepath.Split(fullPath)
+	chunks := strings.Split(dir, remove)
+	if len(chunks) != 2 {
+		return ""
+	}
+	return chunks[1]
+}
+
+func copyFile(dest, src string) error {
```
Comment (krasi-georgiev) [inline fileutil/fileutil.go]: Usually the order is src, dst

File: fileutil/fileutil.go [hunk 4]
```diff
@@ -5,11 +5,102 @@
...
+func getExtraPath(fullPath, remove string) string {
+	// If we have /path/to/file and we want to copy it to
+	// path/copy/to/file, this function should return /copy
+	dir, _ := filepath.Split(fullPath)
+	chunks := strings.Split(dir, remove)
+	if len(chunks) != 2 {
+		return ""
+	}
+	return chunks[1]
+}
```
Comment (krasi-georgiev) [inline fileutil/fileutil.go]: Can you explain why do we need this func, can't understand from the comments.

**Comment by krasi-georgiev** [comment]
> see the latest commit, we also needed to grab extra portions of the path for files we're creating copies of so that the end up in the right destination location

this looks hacky, can you try to simplify.

**Comment by cstyan** [comment]
yes

**Comment by cstyan** [comment]
@krasi-georgiev simplified a bit. We still need a way to remove a prefix from the files we find when using `filepath.Walk`, but we can use `strings.TrimPrefix` rather than the function I had written.

**Comment by krasi-georgiev** [comment]
can you also try the simplified version I suggested for  `copyFile`.

**Comment by cstyan** [comment]
@krasi-georgiev yep, just slipped my mind earlier. 

Want me to squash commits now?

**Comment by krasi-georgiev** [comment]
no need I will squash when I merge it.


moved few bits around and updated few comment.

```
// CopyDirs copies all directories, subdirectories and files recursively including the empty folders.
// Source and destination must be full paths.
func CopyDirs(src, dest string) error {
	if err := os.MkdirAll(dest, 0777); err != nil {
		return err
	}
	files, err := readDirs(src)
	if err != nil {
		return err
	}

	for _, f := range files {
		dp := filepath.Join(dest, f)
		sp := filepath.Join(src, f)

		stat, err := os.Stat(sp)
		if err != nil {
			return err
		}

		// Empty directories are also created.
		if stat.IsDir() {
			if err := os.MkdirAll(dest, 0777); err != nil {
				return err
			}
			continue
		}

		if err := copyFile(sp, dp); err != nil {
			return err
		}
	}
	return nil
}

func copyFile(src, dest string) error {
	data, err := ioutil.ReadFile(src)
	if err != nil {
		return err
	}

	err = ioutil.WriteFile(dest, data, 0644)
	if err != nil {
		return err
	}
	return nil
}

// readDirs reads the source directory recursively
// returns relative paths to all files and empty directories.
func readDirs(src string) ([]string, error) {
	var files []string
	var err error

	err = filepath.Walk(src, func(path string, f os.FileInfo, err error) error {
		relativePath := strings.TrimPrefix(path, src)
		if len(relativePath) > 0 {
			files = append(files, relativePath)
		}
		return nil
	})
	if err != nil {
		return nil, err
	}
	return files, nil
}
```

**Comment by cstyan** [comment]
:+1: changes lgtm, pushed them

**Comment by krasi-georgiev** [comment]
resolve the conflict and LGTM

**Comment by krasi-georgiev** [comment]
Thanks for spending the time on this.

**Pull request merged**